File: mvvmbase/src/main/java/com/bing/mvvmbase/base/BaseViewModel.java
package com.bing.mvvmbase.base;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.LifecycleOwner;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;

public class BaseViewModel extends AndroidViewModel implements DefaultLifecycleObserver {
	protected CompositeDisposable mCompositeDisposable = new CompositeDisposable();
	protected AppExecutors mAppExecutors;

	public BaseViewModel(@NonNull Application application) {
		super(application);
		mAppExecutors = ((BaseApplication) application).getAppExecutors();
	}

	public void addDisposable(Disposable disposable) {
		mCompositeDisposable.add(disposable);
	}

	@Override
	protected void onCleared() {
		mCompositeDisposable.dispose();
		super.onCleared();
	}

        @Override
        public void onCreate(@NonNull LifecycleOwner owner) {
                
        }

        @Override
        public void onResume(@NonNull LifecycleOwner owner) {

        }

        @Override
        public void onStart(@NonNull LifecycleOwner owner) {

        }

        @Override
        public void onPause(@NonNull LifecycleOwner owner) {

        }

        @Override
        public void onStop(@NonNull LifecycleOwner owner) {

        }

        @Override
        public void onDestroy(@NonNull LifecycleOwner owner) {

        }
}

File: mvvmbase/src/main/java/com/bing/mvvmbase/base/BaseViewModel.kt
package com.bing.mvvmbase.base

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.LifecycleOwner
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.disposables.Disposable

open class BaseViewModel(application: Application) : AndroidViewModel(application), DefaultLifecycleObserver {
        protected var mCompositeDisposable = CompositeDisposable()
        protected var mAppExecutors = AppExecutors

        fun addDisposable(disposable: Disposable) {
                mCompositeDisposable.add(disposable)
        }

        override fun onCleared() {
                mCompositeDisposable.dispose()
                super.onCleared()
        }

        override fun onCreate(owner: LifecycleOwner) {

        }

        override fun onResume(owner: LifecycleOwner) {

        }

        override fun onStart(owner: LifecycleOwner) {

        }

        override fun onPause(owner: LifecycleOwner) {

        }

        override fun onStop(owner: LifecycleOwner) {

        }

        override fun onDestroy(owner: LifecycleOwner) {

        }
}
